lib/archive/diff.go

```diff
@@ -0,0 +1,229 @@
+
+	"github.com/vmware/vic/pkg/trace"
+)
+
+const (
+	// ChangeTypeKey defines the key for the type of diff change stored in the tar Xattrs header
+	ChangeTypeKey = "change_type"
+)
+
+// for sort.Sort
+type changesByPath []docker.Change
+
+func (c changesByPath) Less(i, j int) bool { return c[i].Path < c[j].Path }
+func (c changesByPath) Len() int           { return len(c) }
+func (c changesByPath) Swap(i, j int)      { c[j], c[i] = c[i], c[j] }
+
+// Diff produces a tar docker containing the differences between two filesystems
+func Diff(op trace.Operation, newDir, oldDir string, pathspec map[string]string, data bool) (io.ReadCloser, error) {
+
+	changes, err := docker.ChangesDirs(newDir, oldDir)
+	if err != nil {
+		return nil, err
+	}
+
+	sort.Sort(changesByPath(changes))
+
+	spec, err := processPathSpec(op, pathspec)
+	if err != nil {
+		return nil, err
+	}
+
+	// prune excluded paths from change set
+	for i := 0; i < len(changes); {
+		if excluded(op, changes[i].Path, spec) {
+			changes = append(changes[:i], changes[i+1:]...)
+			continue
+		}
+		i++
+	}
+
+	return Tar(op, newDir, changes, spec, data)
+}
+
+func Tar(op trace.Operation, dir string, changes []docker.Change, spec *PathSpec, data bool) (io.ReadCloser, error) {
+	var (
+		err error
+		hdr *tar.Header
+	)
+
+	// Note: it is up to the caller to handle errors and the closing of the read side of the pipe
+	r, w := io.Pipe()
+	go func() {
+		tw := tar.NewWriter(w)
+		defer func() {
+			if cerr := tw.Close(); cerr != nil {
+				op.Errorf("Error closing tar writer: %s", cerr.Error())
+			}
+			if err != nil {
+				_ = w.CloseWithError(err)
+				return
+			}
+			if err = w.Close(); err != nil {
+				op.Errorf("Error closing writer: %s", err.Error())
+			}
+		}()
+
+		for _, change := range changes {
+
+			hdr, err = createHeader(op, dir, change, spec)
+			if err != nil {
+				op.Errorf("Error getting header for change: %s", err.Error())
+				return
+			}
+
+			var f *os.File
+
+			if !data {
+				hdr.Size = 0
+				// include the change type in extended attributes
+				hdr.Xattrs = make(map[string]string)
+				hdr.Xattrs[ChangeTypeKey] = change.Kind.String()
+			}
+
+			_ = tw.WriteHeader(hdr)
+
+			p := filepath.Join(dir, change.Path)
+			if hdr.Typeflag == tar.TypeReg && hdr.Size != 0 {
+				f, err = os.Open(p)
+				if err != nil {
+					if os.IsPermission(err) {
+						err = nil
+					}
+					return
+				}
+				if f != nil {
+					_, err = io.Copy(tw, f)
+					if err != nil {
+						op.Errorf("Error writing docker data: %s", err.Error())
+					}
+					_ = f.Close()
+				}
+
+			}
+
+		}
+	}()
+	return r, err
+}
+
+// grossly inefficient way to handle exclusion/inclusion of paths
+func excluded(op trace.Operation, filePath string, spec *PathSpec) bool {
+	// inefficient way to make sure we don't remove an excluded directory
+	// if it contains a file we explicitly included
+	p := filePath
+	for p != string(filepath.Separator) {
+		if _, ok := spec.Inclusions[p]; ok {
```

Comment:
I know this works well. As you commented that this is one inefficient way, just one suggestion for the improvement:
You can check the exclusion together with inclusion, in the same for loop. 
If that's included, return false as you did.
But if that's excluded, just set a flag excluded = true.
Then after you quit from the loop, return the excluded flag will work same with this.